table/block_based_table_reader.cc

```diff
@@ -217,70 +198,188 @@ bool PrefixExtractorChanged(const TableProperties* table_properties,
 
 }  // namespace
 
+class BlockBasedTable::IndexReaderCommon : public BlockBasedTable::IndexReader {
+public:
+  IndexReaderCommon(BlockBasedTable* t,
+                    CachableEntry<Block>&& index_block)
+      : table_(t)
+      , index_block_(std::move(index_block))
+  {
+    assert(table_ != nullptr);
+  }
+
+protected:
+  static Status ReadIndexBlock(BlockBasedTable* table,
+    FilePrefetchBuffer* prefetch_buffer, const ReadOptions& read_options,
+    GetContext* get_context, CachableEntry<Block>* index_block);
+
+  BlockBasedTable* table() const { return table_; }
+
+  const InternalKeyComparator* internal_comparator() const {
+    assert(table_ != nullptr);
+    assert(table_->get_rep() != nullptr);
+
+    return &table_->get_rep()->internal_comparator;
+  }
+
+  bool index_key_includes_seq() const {
+    assert(table_ != nullptr);
+    assert(table_->get_rep() != nullptr);
+
+    const TableProperties* const properties =
+      table_->get_rep()->table_properties.get();
+
+    return properties == nullptr || !properties->index_key_is_user_key;
+  }
+
+  bool index_value_is_full() const {
+    assert(table_ != nullptr);
+    assert(table_->get_rep() != nullptr);
+
+    const TableProperties* const properties =
+      table_->get_rep()->table_properties.get();
+
+    return properties == nullptr || !properties->index_value_is_delta_encoded;
+  }
+
+  Status GetIndexBlock(const ReadOptions& read_options,
+                       GetContext* get_context,
+                       CachableEntry<Block>* index_block) const;
+
+  size_t ApproximateIndexBlockMemoryUsage() const {
+    assert(!index_block_.GetOwnValue() || index_block_.GetValue() != nullptr);
+    return index_block_.GetOwnValue() ?
+      index_block_.GetValue()->ApproximateMemoryUsage() : 0;
+  }
+
+private:
+  BlockBasedTable* table_;
+  CachableEntry<Block> index_block_;
+};
+
+Status BlockBasedTable::IndexReaderCommon::ReadIndexBlock(
+  BlockBasedTable* table, FilePrefetchBuffer* prefetch_buffer,
+  const ReadOptions& read_options, GetContext* get_context,
+  CachableEntry<Block>* index_block) {
+
+  PERF_TIMER_GUARD(read_index_block_nanos);
+
+  assert(table != nullptr);
+  assert(index_block != nullptr);
+  assert(index_block->IsEmpty());
+
+  const Rep* const rep = table->get_rep();
+  assert(rep != nullptr);
+
+  constexpr bool is_index = true;
+  const Status s = BlockBasedTable::RetrieveBlock(prefetch_buffer,
+    rep, read_options, rep->footer.index_handle(),
+    UncompressionDict::GetEmptyDict(), index_block, is_index, get_context);
+
+  return s;
+}
+
+Status BlockBasedTable::IndexReaderCommon::GetIndexBlock(
```

Comment:
Its not clear why we need both ```ReadIndexBlock``` and ```GetIndexBlock```. Can't they be combined?

Reply:
ReadIndexBlock retrieves the index block from the block cache or the file itself.

GetIndexBlock first checks if the reader already has the index block in the index_block_ member variable (this would be the case if cache_index_and_filter_blocks==false, or if it's true AND pinning is enabled), and falls back to ReadIndexBlock if it's not.

In addition to being used inside GetIndexBlock, ReadIndexBlock is also called (directly) when the IndexReader is first created.

Reply:
Renamed GetIndexBlock to GetOrReadIndexBlock to make this clearer.